File: apps/web/lib/api/matches.ts
/**
 * API client for match management.
 */

import api from './index'

// =============================================================================
// Types
// =============================================================================

export interface MatchCreate {
    case_id: string
    intended_parent_id: string
    compatibility_score?: number
    notes?: string
}

export interface MatchRead {
    id: string
    case_id: string
    intended_parent_id: string
    status: string
    compatibility_score: number | null
    proposed_by_user_id: string | null
    proposed_at: string
    reviewed_by_user_id: string | null
    reviewed_at: string | null
    notes: string | null
    rejection_reason: string | null
    created_at: string
    updated_at: string
    case_number: string | null
    case_name: string | null
    ip_name: string | null
    // Case stage info for status sync
    case_stage_id: string | null
    case_stage_slug: string | null
    case_stage_label: string | null
}

export interface MatchListItem {
    id: string
    case_id: string
    case_number: string | null
    case_name: string | null
    intended_parent_id: string
    ip_name: string | null
    status: string
    compatibility_score: number | null
    proposed_at: string
    // Case stage info for status sync
    case_stage_id: string | null
    case_stage_slug: string | null
    case_stage_label: string | null
}

export interface MatchListResponse {
    items: MatchListItem[]
    total: number
    page: number
    per_page: number
}

export interface MatchStatsResponse {
    total: number
    by_status: Record<MatchStatus, number>
}

export interface MatchAcceptRequest {
    notes?: string
}

export interface MatchRejectRequest {
    notes?: string
    rejection_reason: string
}

export interface MatchUpdateNotesRequest {
    notes: string
}

export type MatchStatus = 'proposed' | 'reviewing' | 'accepted' | 'rejected' | 'cancelled'

// =============================================================================
// API Functions
// =============================================================================

export interface ListMatchesParams {
    status?: MatchStatus
    case_id?: string
    intended_parent_id?: string
    page?: number
    per_page?: number
}

/**
 * List matches with optional filters.
 */
export async function listMatches(params: ListMatchesParams = {}): Promise<MatchListResponse> {
    const searchParams = new URLSearchParams()
    if (params.status) searchParams.set('status', params.status)
    if (params.case_id) searchParams.set('case_id', params.case_id)
    if (params.intended_parent_id) searchParams.set('intended_parent_id', params.intended_parent_id)
    if (params.page) searchParams.set('page', params.page.toString())
    if (params.per_page) searchParams.set('per_page', params.per_page.toString())
    const query = searchParams.toString()
    return api.get<MatchListResponse>(`/matches/${query ? `?${query}` : ''}`)
}

/**
 * Get match counts by status.
 */
export async function getMatchStats(): Promise<MatchStatsResponse> {
    return api.get<MatchStatsResponse>('/matches/stats')
}

/**
 * Get match by ID.
 */
export async function getMatch(matchId: string): Promise<MatchRead> {
    return api.get<MatchRead>(`/matches/${matchId}`)
}

/**
 * Create a new match proposal.
 */
export async function createMatch(data: MatchCreate): Promise<MatchRead> {
    return api.post<MatchRead>('/matches/', data)
}

/**
 * Accept a match.
 */
export async function acceptMatch(matchId: string, data: MatchAcceptRequest = {}): Promise<MatchRead> {
    return api.put<MatchRead>(`/matches/${matchId}/accept`, data)
}

/**
 * Reject a match with reason.
 */
export async function rejectMatch(matchId: string, data: MatchRejectRequest): Promise<MatchRead> {
    return api.put<MatchRead>(`/matches/${matchId}/reject`, data)
}

/**
 * Cancel a proposed match.
 */
export async function cancelMatch(matchId: string): Promise<void> {
    await api.delete(`/matches/${matchId}`)
}

/**
 * Update match notes.
 */
export async function updateMatchNotes(matchId: string, data: MatchUpdateNotesRequest): Promise<MatchRead> {
    return api.patch<MatchRead>(`/matches/${matchId}/notes`, data)
}

// =============================================================================
// Match Events Types
// =============================================================================

export type MatchEventPersonType = 'surrogate' | 'ip'
export type MatchEventType = 'medication' | 'medical_exam' | 'legal' | 'delivery' | 'custom'

export interface MatchEventCreate {
    person_type: MatchEventPersonType
    event_type: MatchEventType
    title: string
    description?: string | null
    starts_at?: string | null
    ends_at?: string | null
    timezone?: string
    all_day?: boolean
    start_date?: string | null
    end_date?: string | null
}

export interface MatchEventUpdate extends Partial<MatchEventCreate> { }

export interface MatchEventRead {
    id: string
    match_id: string
    person_type: MatchEventPersonType
    event_type: MatchEventType
    title: string
    description: string | null
    starts_at: string | null
    ends_at: string | null
    timezone: string
    all_day: boolean
    start_date: string | null
    end_date: string | null
    created_by_user_id: string | null
    created_at: string
    updated_at: string
}

// =============================================================================
// Match Events API Functions
// =============================================================================

/**
 * List events for a match.
 */
export async function listMatchEvents(matchId: string): Promise<MatchEventRead[]> {
    return api.get<MatchEventRead[]>(`/matches/${matchId}/events`)
}

/**
 * Get a specific match event.
 */
export async function getMatchEvent(matchId: string, eventId: string): Promise<MatchEventRead> {
    return api.get<MatchEventRead>(`/matches/${matchId}/events/${eventId}`)
}

/**
 * Create a match event.
 */
export async function createMatchEvent(matchId: string, data: MatchEventCreate): Promise<MatchEventRead> {
    return api.post<MatchEventRead>(`/matches/${matchId}/events`, data)
}

/**
 * Update a match event.
 */
export async function updateMatchEvent(matchId: string, eventId: string, data: MatchEventUpdate): Promise<MatchEventRead> {
    return api.put<MatchEventRead>(`/matches/${matchId}/events/${eventId}`, data)
}

/**
 * Delete a match event.
 */
export async function deleteMatchEvent(matchId: string, eventId: string): Promise<void> {
    await api.delete(`/matches/${matchId}/events/${eventId}`)
}

// =============================================================================
// Color Utilities
// =============================================================================

export const EVENT_TYPE_COLORS: Record<MatchEventType, string> = {
    medication: '#f97316',
    medical_exam: '#3b82f6',
    legal: '#eab308',
    delivery: '#ef4444',
    custom: '#6b7280',
}

export const PERSON_TYPE_COLORS: Record<MatchEventPersonType, string> = {
    surrogate: '#a855f7',
    ip: '#22c55e',
}

File: apps/api/app/schemas/matches.py
"""Match API schemas."""

from datetime import date as date_type
from datetime import datetime
from typing import Literal
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field, field_validator, model_validator

from app.db.enums import MatchStatus


class MatchCreate(BaseModel):
    """Request to propose a match."""

    surrogate_id: UUID | None = None
    donor_id: UUID | None = None
    intended_parent_id: UUID
    notes: str | None = None

    @model_validator(mode="after")
    def validate_participants(self):
        if (self.surrogate_id is None) == (self.donor_id is None):
            raise ValueError("Exactly one surrogate or donor is required")
        return self


class MatchRead(BaseModel):
    """Match response."""

    id: str
    match_number: str
    surrogate_id: str | None
    donor_id: str | None = None
    donor_name: str | None = None
    donor_number: str | None = None
    donor_stage_label: str | None = None
    match_kind: Literal["surrogate", "donor"] = "surrogate"
    closed_at: str | None = None
    closure_reason: str | None = None
    outcome: str | None = None
    intended_parent_id: str
    status: MatchStatus
    proposed_by_user_id: str | None
    proposed_at: str
    reviewed_by_user_id: str | None
    reviewed_at: str | None
    notes: str | None
    decline_reason: str | None
    created_at: str
    updated_at: str
    # Denormalized for convenience
    surrogate_number: str | None = None
    surrogate_name: str | None = None
    ip_name: str | None = None
    ip_number: str | None = None
    # Surrogate stage info for status sync
    surrogate_stage_id: str | None = None
    surrogate_stage_slug: str | None = None
    surrogate_stage_label: str | None = None


class MatchListItem(BaseModel):
    """Match list item with summary info."""

    id: str
    match_number: str
    surrogate_id: str | None
    donor_id: str | None = None
    donor_name: str | None = None
    donor_number: str | None = None
    donor_stage_label: str | None = None
    match_kind: Literal["surrogate", "donor"] = "surrogate"
    closed_at: str | None = None
    closure_reason: str | None = None
    outcome: str | None = None
    surrogate_number: str | None
    surrogate_name: str | None
    intended_parent_id: str
    ip_name: str | None
    ip_number: str | None = None
    status: MatchStatus
    proposed_at: str
    # Surrogate stage info for status sync
    surrogate_stage_id: str | None = None
    surrogate_stage_slug: str | None = None
    surrogate_stage_label: str | None = None


class MatchListResponse(BaseModel):
    """Paginated match list."""

    items: list[MatchListItem]
    total: int
    page: int
    per_page: int


class MatchStatsResponse(BaseModel):
    """Match stats summary."""

    total: int
    by_status: dict[MatchStatus, int]


class MatchAcceptRequest(BaseModel):
    """Request to accept a match."""

    notes: str | None = None


class MatchDeclineRequest(BaseModel):
    """Request to decline a match."""

    notes: str | None = None
    reason: str = Field(min_length=1)

    @field_validator("reason")
    @classmethod
    def required_reason(cls, value: str) -> str:
        if not value.strip():
            raise ValueError("Decline reason is required")
        return value.strip()


class MatchCancelRequest(BaseModel):
    """Request to cancel an accepted match (admin approval required)."""

    reason: str = Field(min_length=1)

    @field_validator("reason")
    @classmethod
    def required_reason(cls, value: str) -> str:
        if not value.strip():
            raise ValueError("Cancellation reason is required")
        return value.strip()


class MatchUpdateNotesRequest(BaseModel):
    """Request to update match notes."""

    notes: str


class MatchEventCreate(BaseModel):
    """Request to create a match event."""

    person_type: str = Field(pattern="^(surrogate|donor|ip)$")
    event_type: str = Field(pattern="^(medication|medical_exam|legal|delivery|custom)$")
    title: str = Field(min_length=1, max_length=200)
    description: str | None = None
    starts_at: datetime | None = None
    ends_at: datetime | None = None
    timezone: str = "America/Los_Angeles"
    all_day: bool = False
    start_date: str | None = None  # YYYY-MM-DD for all-day events
    end_date: str | None = None


class MatchEventUpdate(BaseModel):
    """Request to update a match event."""

    person_type: str | None = Field(None, pattern="^(surrogate|donor|ip)$")
    event_type: str | None = Field(
        None, pattern="^(medication|medical_exam|legal|delivery|custom)$"
    )
    title: str | None = Field(None, min_length=1, max_length=200)
    description: str | None = None
    starts_at: datetime | None = None
    ends_at: datetime | None = None
    timezone: str | None = None
    all_day: bool | None = None
    start_date: str | None = None
    end_date: str | None = None


class MatchEventRead(BaseModel):
    """Match event response."""

    id: str
    match_id: str
    person_type: str
    event_type: str
    title: str
    description: str | None
    starts_at: str | None
    ends_at: str | None
    timezone: str
    all_day: bool
    start_date: str | None
    end_date: str | None
    created_by_user_id: str | None
    created_at: str
    updated_at: str


class MatchCompleteRequest(BaseModel):
    outcome: str = Field(min_length=1, max_length=2000)
    reason: str | None = Field(None, max_length=2000)


class AttemptCreate(BaseModel):
    attempt_type: Literal["embryo_transfer", "retrieval", "collection", "other"]
    status: Literal["planned", "in_progress", "completed", "cancelled"] = "planned"
    started_at: date_type | None = None
    ended_at: date_type | None = None
    outcome: str | None = Field(None, max_length=2000)


class AttemptUpdate(BaseModel):
    attempt_type: Literal["embryo_transfer", "retrieval", "collection", "other"] | None = None
    status: Literal["planned", "in_progress", "completed", "cancelled"] | None = None
    started_at: date_type | None = None
    ended_at: date_type | None = None
    outcome: str | None = Field(None, max_length=2000)

    @model_validator(mode="after")
    def reject_null_required(self):
        if any(
            field in self.model_fields_set and getattr(self, field) is None
            for field in ("attempt_type", "status")
        ):
            raise ValueError("Attempt type and status cannot be null")
        return self


class AttemptRead(AttemptCreate):
    model_config = ConfigDict(from_attributes=True)
    id: UUID
    match_id: UUID
    sequence: int
    created_at: datetime
    updated_at: datetime
